### PR TITLE
wip: support test_ api key

### PR DESCRIPTION
**Pull request by buwilliams**
## Changes

- Support for `test_[apiKey]` which signifies local, test, and staging environments
- Allows for filtering all data by drawing a line between Production and everything other env
- `test_[apiKey]` sets `$environment` to `test`
- Users may also pass in this property to override or simply set the value for `$environment`
- Added environment toggle to top navigation and put it behind a feature flag

See #3149 for more information

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [x] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [x] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious

### CODE REVIEW

File: posthog/api/capture.py
Comment (timgl) [inline posthog/api/capture.py:245]: Should probably a property with a `$` to indicate that it's a standard posthog one? Then you can also add it to PropertyKeyInfo.tsx. Also worth adding to `posthog/models/team.py` so it's included in the standard filter.

Comment (buwilliams) [reply posthog/api/capture.py:245]: I'll add the `$`. I can it to team/propertykey but should we? Currently, it's meant to be hidden and not selectable in the UI.

Comment (buwilliams) [reply posthog/api/capture.py:245]: I've been looking at `posthog/models/team.py` but I'm unable to figure out what you had in mind?

File: frontend/src/lib/components/PropertyFilters/PropertyFilters.js
Comment (timgl) [inline frontend/src/lib/components/PropertyFilters/PropertyFilters.js:98]: I'd suggest using the data in PropertyKeyInfo.tsx here instead of this random array

File: posthog/api/capture.py
Comment (timgl) [reply posthog/api/capture.py]: re propertykeyinfo, you can `hide: true` to hide it in the event view etc

re: teams, I'd suggest adding a condition to the test filters (https://github.com/PostHog/posthog/blob/master/posthog/models/team.py#L36) so that when someone ticks 'filter test accounts' these events get filtered too.

Comment (buwilliams) [reply posthog/api/capture.py]: I think there are differences between the way you are thinking about this and the way Paolo and I were thinking about it. I'm struggling to identify exactly what that difference is. I think it has something to do with "filter test accounts" since this feature supersedes that feature. It's not meant to be integrated into the existing filters. It's a matter of the state being managed outside the regular filters. I'll give it some more thought because you may be right about this.

**Comment by buwilliams** [comment]
Anyone know what this check is failing? It's a python issue that doesn't make sense to me.

File: posthog/api/capture.py
Comment (paolodamico) [reply posthog/api/capture.py]: So actually @buwilliams, I think we left an asterisk on what to do with the existing test filters. I'm strongly leaning on keeping it as an advanced feature both for backwards-compatibility and support other use cases where changing the event data is not the best user experience.

In this case, the test account filters switch (frontend) would disappear as it is today, and when the new general switch is toggled to test, we would include the internal/test filters too.

From a backend standpoint it might actually be a good way to go to rely on the existing test filters logic to filter out these events instead of adding the `test_environment` property to each request (as Tim is suggesting). Thoughts?

**Comment by paolodamico** [comment]
Re @buwilliams,
> Anyone know what this check is failing? It's a python issue that doesn't make sense to me.

The check that was failing indicated that there was a typing problem. Python has only limited support for typing, introduced fairly recently ([see more details](https://docs.python.org/3/library/typing.html)). One of the limitations is that typing is not actually checked when coding (some IDEs might support this) or at runtime. Instead, we use https://github.com/python/mypy to check typing. You can run it yourself locally by doing `DEBUG=1 mypy .`

See 87018fca90cde75ea0f0466c7b7dd6346e1c5dfe for details on how the specific typing issue was fixed.

File: frontend/src/scenes/userLogic.tsx
Comment (paolodamico) [inline frontend/src/scenes/userLogic.tsx:44]: I think it's better to keep this logic in `navigationLogic`, as it's related to those components and not user-scoped. I would also not treat this as a boolean state but set the two options explicitly (for code clarity purposes, it also avoids confusion on later implementations). Something like: `type GlobalEnvironment = 'live' | 'test'`

File: frontend/src/lib/components/PropertyKeyInfo.tsx
Comment (paolodamico) [inline frontend/src/lib/components/PropertyKeyInfo.tsx:436]: I think @timgl mentioned you can actually just use the existing logic for this, there's a `hide` property (see line 281) that handles this behavior.

**Review by paolodamico — CHANGES_REQUESTED** [review]
Added some comments. Also proposing a different UI for the toggle (feel free to challenge / iterate).

1. Also worth noting, what happens when you change the filter? Ideally we would reload the graph, table you're viewing.
2. Finally, we should also remove the test filters toggle from the UI if the feature flag is enabled.

**Comment by paolodamico** [comment]
Here's my proposed UI for this. Feedback @corywatilo too? 

image


## Outstanding

Please note I think we should also add a global indicator to all pages to be more explicit when you're on test data mode. Something like this,

**Review by Twixes — CHANGES_REQUESTED** [review]
Hey! So, I posted the original env separation issue, and I think this implementation may be too narrow. I think it's quite possible we'll want to support more such environments than with just a boolean switch (e.g. selector a la Sentry), so it'd be more flexible if instead of `$test_environment: true/false` we used `$environment: 'production'/'development'`.
Aaand, that still does not address a key issue that motivates this feature request: users want to get rid of environments with junk data, but that's not possible - events, once ingested, they history forever, because we don't delete CH events yet. With projects that's _kind of_ solved because the data is just inaccessible once the project is deleted, thanks to using `team_id` for all filtering.

File: posthog/api/capture.py
Comment (buwilliams) [reply posthog/api/capture.py]: @paolodamico adding the `test_environment` property may be necessary since the current "filter test data" assumes the data needed to filter is present. Otherwise, how do you filter for test environments if we currently are not tracking that an environment is a test environment? Additionally, the case was for non-technical users who by adding `test_` to their apiKey to get this functionality for free. Adding test_[apiKey] allows us to inject that information so that the filtering test data has a property to use. I think that is a nice user experience otherwise users will need to manually add `test_environment` to all their capture calls. URLs won't work since we support more than browsers.

re: "advanced feature both for backwards-compatibility and support other use cases where changing the event data is not the best user experience" - What is an advanced feature? What makes an advanced feature backwards compatible? Why does change event data not a good user experience? I could see only adding the test_environment in the test_ case and not all cases but I'm unsure if that is what you mean.

re: exist test filters logic - Leveraging existing test filters logic seems good to me.

**Comment by buwilliams** [comment]
@Twixes Why does removing the data matter? Filtering removes unwanted results so the effect is the same from a UI perspective. Additionally, if a flag such as test_environment is present, deleting that data becomes trivial. In either case, removing the data seems like a follow-up issue. One that allows users to purge data based on a filter.

The reason for having boolean rather than by environment was to aid UX since we can think about production and all others and two distinct things and present a simple UX for it. Otherwise, we'd need to create an additional table where users would be to specify their environments so that we could group them appropriately. Seems to me, if an advanced user wanted to track per environment, they could add this as a regular event property. Something like: `environment=production`

File: frontend/src/lib/components/PropertyKeyInfo.tsx
Comment (buwilliams) [reply frontend/src/lib/components/PropertyKeyInfo.tsx:436]: Makes sense to me.

**Comment by buwilliams** [comment]
I've changed the Boolean to a string and renamed it to $environment. Currently, it supports: "production" and "test". Also, I've moved it into a hidden property. Push coming soon.

**Comment by Twixes** [comment]
Yep, okay, actually removing the data from our disks doesn't matter (that much), but removing it from the user's point of view (meaning just making it completely inaccessible) matters. So it's enough to just always and everywhere filter based on this property. And the environments for a project can be just a Postgres array of strings on the `Team` model for now.
BTW if we introduce this, it'd be better as an actual Postgres/ClickHouse column on the events table – getting the property from JSON is significantly more expensive. Deployments & Infrastructure would be of help.

**Comment by paolodamico** [comment]
I agree with @buwilliams, I think multiple environments could be a more advanced use and to consider for later.

In terms of data deletion, I'm torn but leaning towards this is the right approach. Deleting data is a separate thing we need to tackle. 

@buwilliams perhaps it's useful to add the context on why we decided against separate projects for this ?

**Comment by buwilliams** [comment]
@paolodamico @timgl @Twixes before I continue down this road, I've made a bunch of changes so could you guys take a look and give me your thoughts?

@Twixes I decided against multiple projects because:

1. If we created a new project for each environment we *change the meaning of project* when in reality a project shouldn't have anything to do with an environment. Multiple projects means more setup overhead anytime a new project is created. Imagine a user who has multiple projects with multiple environments. It's better to support environments within a project.
2. Separate projects means the data is siloed so it's difficult to ask questions of all test environments.
3. All filters and settings are shared in a single project so anything the user creates will remain across environments which is a nice UX.
4. Choosing to go with production and all other environments makes the UI/UX a binary choice. Do I want to see test or production data? This is just conceptually nice. Additionally, it's always possible to filter by specific environment with normal properties for advanced users.
5. Add to that the cost of implementation and we landed on a simpler solution. There's a lot of ways to solve for this so I choose the option which gave us the greatest flexibility to change our minds since we do not have multiple projects to deal with.

**Comment by corywatilo** [comment]
@paolodamico Looks like I'm a little late to the party, but I think that works. Definitely like the orange line indicating test mode.

One way to simplify would be to only use the word of the current state to the left of the toggle:

image

**Comment by buwilliams** [comment]
@corywatilo That was one of the first UX we tried. It was unintuitive to me that clicking test mode would change the state to live. Having the words on the left and right at least made it clear that you were toggling between states rather than turning something on or off.

**Comment by corywatilo** [comment]
Ya I can see that. Stripe has it _live by default_ and you can toggle _on_ test mode, which makes a bit more sense.
image

**Comment by buwilliams** [comment]
@corywatilo makes sense to me. @paolodamico any objections?

File: posthog/models/event.py
Comment (paolodamico) [inline posthog/models/event.py:34]: Nit: Not sure if this would be better in constants.py

File: frontend/src/lib/components/PropertyKeyInfo.tsx
Comment (paolodamico) [inline frontend/src/lib/components/PropertyKeyInfo.tsx:442]: Think this might be better as an `enum`? I would also keep this in `constants.tsx` as it's shared across files

File: frontend/src/layout/navigation/navigationLogic.ts
Comment (paolodamico) [inline frontend/src/layout/navigation/navigationLogic.ts:73]: Nit: You can destructure the payload here to follow convention: `setFilteredEnvironment: (_, { environment } ) => environment`

File: frontend/src/layout/navigation/navigationLogic.ts
Comment (paolodamico) [inline frontend/src/layout/navigation/navigationLogic.ts:31]: type of `environment` variable should be the enum/const options, not just any string.

**Comment by paolodamico** [comment]
I like it @corywatilo / @buwilliams ! How's this?

image

**Comment by corywatilo** [comment]
love #shipit

**Comment by buwilliams** [comment]
Toggle is working. This "project" (not PR) still needs work but I think we should merge soon. Future changes will be somewhat large. Tomorrow I will incorporate @paolodamico 's feedback and I think we can merge.

Future changes:
- Remove localStorage by instantiating Api() along with React.
- Remove location.reload() the by refreshing logics on environment toggle.
- Full UI/Backend for all environments, defaults: test, staging, production (with a column boolean for prod)

Update: Found a bug preventing incoming events to fix before merge.

**Comment by buwilliams** [comment]
I'm blocked until M1. What I thought was a bug in my code was not. Events are not ingested because we process events by sending them to a [job queue on our plugin server](https://github.com/PostHog/posthog/blob/master/posthog/api/capture.py#L258). I was misguided because `/demo` inserts history for us which appeared as though I was receiving events. Our [plugin server relies on Clickhouse](https://github.com/PostHog/plugin-server/blob/master/package.json#L37) which isn't support on M1 as most of you know. Maybe someone knows a way around this but I haven't been able to find it.

**Comment by mariusandra** [comment]
To clarify one point, the plugin server definitely does not rely on clickhouse, as we also use the same plugin server for the OSS / non-clickhouse installations. Without explicitly setting `KAFKA_ENABLED` to true, the plugin server uses the standard postgres ingestion.

**Comment by buwilliams** [comment]
Solved the issue with the plugin-server by running `yarn build; WORKER_CONCURRENCY=2 yarn start:dist` and created an [issue #370](https://github.com/PostHog/plugin-server/issues/370). This is ready for merge once all the checks pass, we can continue to improve it in another PR.

**Pull request merged**
